serial/pet/driver.s

```diff
@@ -0,0 +1,730 @@
+                ; Shift pressed, read upper table
+                lda KbdMatrixShift,X  ;4;
+                rts                   ;6; (57/59 cycles to here)
+
+@notshift:      lda KbdMatrix,X       ;4;
+                bmi @special          ;2/3; Don't have control modify special keys
+                lda CtrlFlag          ;3;
+                beq @notctrl          ;2/3
+                ; Ctrl pressed, read lower table and bitmask to CtrlFlag keys
+                lda KbdMatrix,X       ;4;
+                and #$9F              ;2;
+@special:       rts                   ;6; (71/73 if we didn't take @special)
+                                      ;   (61/63 if we took @special)
+                ; Normal key
+@notctrl:       lda KbdMatrix,X       ;4;
+                rts                   ;6; (71/73 to here)
+
+@k7:            lda #0                ;2; Table is backwards
+                beq @found            ;3;
+@k6:            lda #1                ;2;
+                bne @found            ;3;
+@nokey:         rts                   ;6; (20 cycles to here)
+
+.endif          ; Baud-based keyboard polling routines
+
+;-----------------------------------------------------------------------------------
+; IRQ handler that is installed when OpenSerial is called
+;-----------------------------------------------------------------------------------
+
+IrqHandler:     ; 36 cycles till we hit here from IRQ firing
+                ; 3 possible interrupt sources: (order of priority)
+                ;  TIM2 - RX timer (after start bit)
+                ;  CA1 falling - Start bit of data to recieve
+                ;  TIM1 - TX timer/kbd poll
+
+                lda #$20              ;2; TIMER2 flag
+                bit VIA_IFR           ;4;
+                bne @tim2             ;3; CA1 triggered $02
+                bvs @tim1             ; Timer 1       $40
+                jmp @ca1
+
+                ; Timer 2  $20
+@tim2:          lda VIA_T2CL          ;4; Acknowledge
+                lda VIA_PA1           ;4; Clear any pending CA1 interrupts
+                ; Read in bit from serial port, build up byte
+                ; If 8 recieved, indicate byte, and disable our
+                ; interrupt
+                lda VIA_PA2           ;4;
+                and #$01              ;2; Only read the Rx pin
+                ror                   ;2; Move into carry
+                ror RxCurByte         ;5;
+
+                dec RxBitNum          ;5;
+                bne @tim2retrig       ;3;
+
+                ; We've receieved a byte, signal to program
+                ; disable our interrupt
+
+                ldx RxBufWritePtr
+                lda RxCurByte
+                sta RX_BUF,X
+
+                inc RxBufWritePtr
+
+                lda #$22              ; Disable timer 2 interrupt and CA1
+                sta VIA_IER
+                lda #$82              ; Enable CA1 interrupt
+                sta VIA_IER
```

Comment:
Same here again, just confirming that intention is to his the same register twice

Reply:
Yes, the first write disables two interrupts (bit 7 is 0), the second enables one (bit 7 is 1).